presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanChecker.java

```diff
@@ -77,26 +91,95 @@ public PlanChecker(FeaturesConfig featuresConfig, boolean forceSingleNode)
+    public void validatePlanFragment(PlanFragment planFragment, Session session, Metadata metadata, WarningCollector warningCollector)
+    {
+        ensureProvidedInitialized();
+        checkers.get(Stage.FRAGMENT).forEach(checker -> checker.validateFragment(planFragment, session, metadata, warningCollector));
+    }
+
+    private static List<Checker> fromSpi(List<com.facebook.presto.spi.plan.PlanChecker> checkers)
+    {
+        return checkers.stream().map(SpiCheckerAdapter::new).collect(Collectors.toList());
+    }
+
+    private void ensureProvidedInitialized()
     {
-        checkers.get(Stage.FRAGMENT).forEach(checker -> checker.validate(planNode, session, metadata, warningCollector));
+        if (!providedInitialized) {
+            if (providerMgr != null) {
+                checkers = ImmutableListMultimap.<Stage, Checker>builder().putAll(checkers)
+                        .putAll(
+                                Stage.INTERMEDIATE,
+                                fromSpi(providerMgr.getPlanCheckerProvider().getPlanCheckersIntermediate()))
+                        .putAll(
+                                Stage.FRAGMENT,
+                                fromSpi(providerMgr.getPlanCheckerProvider().getPlanCheckersFragment()))
+                        .putAll(
+                                Stage.FINAL,
+                                fromSpi(providerMgr.getPlanCheckerProvider().getPlanCheckersFinal()))
+                        .build();
+            }
+            providedInitialized = true;
+        }
     }
 
     public interface Checker
     {
         void validate(PlanNode planNode, Session session, Metadata metadata, WarningCollector warningCollector);
+
+        default void validateFragment(PlanFragment planFragment, Session session, Metadata metadata, WarningCollector warningCollector)
+        {
+            validate(planFragment.getRoot(), session, metadata, warningCollector);
+        }
     }
 
     private enum Stage
     {
         INTERMEDIATE, FINAL, FRAGMENT
     }
+
+    private static class SpiCheckerAdapter
+            implements Checker
+    {
+        private final com.facebook.presto.spi.plan.PlanChecker checker;
+
+        public SpiCheckerAdapter(com.facebook.presto.spi.plan.PlanChecker checker)
+        {
+            this.checker = checker;
+        }
+
+        @Override
+        public void validate(PlanNode planNode, Session session, Metadata metadata, WarningCollector warningCollector)
+        {
+            checker.validate(planNode, warningCollector);
+        }
+
+        @Override
+        public void validateFragment(PlanFragment planFragment, Session session, Metadata metadata, WarningCollector warningCollector)
+        {
+            checker.validateFragment(toSimplePlanFragment(planFragment), warningCollector);
+        }
+
+        private SimplePlanFragment toSimplePlanFragment(PlanFragment planFragment)
```

Comment:
This can be static, but maybe instead put this inside PlanFragment and make it an object method on a PlanFragment (so you would call planFragment.toSimplePlanFragment()).

Reply:
good point, it's only used in validatePlanFragment so I moved it there